Nodejs/Product/Nodejs/TypingsAcquisition.cs

```diff
@@ -0,0 +1,143 @@
+﻿//*********************************************************//
+//    Copyright (c) Microsoft. All rights reserved.
+//    
+//    Apache 2.0 License
+//    
+//    You may obtain a copy of the License at
+//    http://www.apache.org/licenses/LICENSE-2.0
+//    
+//    Unless required by applicable law or agreed to in writing, software 
+//    distributed under the License is distributed on an "AS IS" BASIS, 
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or 
+//    implied. See the License for the specific language governing 
+//    permissions and limitations under the License.
+//
+//*********************************************************//
+
+using System.Threading.Tasks;
+using Microsoft.VisualStudioTools.Project;
+using Microsoft.NodejsTools.Npm;
+using System.Collections.Generic;
+using System.Linq;
+using System.IO;
+
+using SR = Microsoft.NodejsTools.Project.SR;
+using System;
+using System.Security;
+
+namespace Microsoft.NodejsTools {
+    internal class TypingsAcquisition {
+        private const string TsdExe = "tsd.cmd";
+        private const string TypingsDirectoryName = "typings";
+
+        private readonly string _pathToRootNpmDirectory;
+        private readonly string _pathToRootProjectDirectory;
+        private readonly Lazy<HashSet<string>> _acquiredTypingsPackageNames;
+
+        public TypingsAcquisition(
+            string pathToRootNpmDirectory,
+            string pathToRootProjectDirectory) {
+            _pathToRootNpmDirectory = pathToRootNpmDirectory;
+            _pathToRootProjectDirectory = pathToRootProjectDirectory;
+
+            _acquiredTypingsPackageNames = new Lazy<HashSet<string>>(() => {
+                return new HashSet<string>(CurrentTypingsPackages(_pathToRootProjectDirectory));
+            });
+        }
```

Comment:
Are you proposing that we have a top level `typings` directory instead of one in each project's root? For example, with projects A & B, the current setup is:

```
A/
    ...
    package.json
    typings/...
B/
    ...
    package.json
    typings
```

With the singleton, are you instead proposing:

```
typings/...
A
    ...
    package.json
B
    ...
    package.json
```